### DISCUSSION TITLE
Adding More Scan Lists - Now 10 lists

### DISCUSSION THREAD

**Discussion by AubsUK** [egzumer/uv-k5-firmware-custom#480] [Show and tell]
I originally raised **AND CLOSED** PR [#421 - AubsUK - Added Scan option to scan List1, List2, All Channels, All Lists, No Lists](https://github.com/egzumer/uv-k5-firmware-custom/pull/421) to add additional flexibility to how the UV-K5 works with scanning, to stretch the number of Lists available.  @qsus asked in the PR thread about adding more lists, and I've finally been able to do it.  It's rough at the moment and will need more work to get it perfect, but the basis is there.  I would love it to be included in the Egzumer firmware.
<br/><br/>
### Help!
I would appreciate initial thoughts before I spend much more time on it - Keep Reading!

I would also appreciate any advice with getting the Chirp driver configured appropriately.

One question - Would it be better to have it as an optional feature, or keep it embedded?
<br/><br/>
### Compiled Size
EGZUMER v0.21 -	Firmware uses 99.04% of available memory (60852/61439 bytes).
My latest version -	Firmware uses 99.93% of available memory (61396/61439 bytes).
61396-60852=544 bytes extra used.
<br/><br/>
### To Do
I know I need to re-write a bit of the code to make it easier to handle, so I've not shared the code just yet:
 - I've shortened the memory space used for the Channel Name (e.g. for MR Channel 1 Name from `0x0F50..0x0F5F` to `0x0F50.. 0x0F5D`) and used remaining two bytes (e.g. for MR Channel 1 Lists this is `0x0F5E..0x0F5F`).  I'll probably re-write it as a `struct` containing both (I didn't know struct's existed when I started this!!!).  This will probably reduce the compiled size a bit too
 - The existing memory space previously used is all set to 1's (`memset(Template, 0xFF, sizeof(Template));`), so that messes up any unused channels at the moment (I've baked in a way to manually clear them all down, but I suspect the above option will help with this).
 - I've removed priority watch channels from the code as they weren't implemented anyway, and took up space unnecessarily.
 - I want to add in an option to Auto-Scan on power on option (I've already written it into the saved lists).
 - I disabled the 'DW on hold' symbol `><` that showed when scanning lists.
<br/><br/>
### What can it do

- There are 10 lists, 0,1,2,3,4,5,6,7,8,9.
- Each list can be turned on or off independently (written to memory so rebooting retains the list)
- Every channel can be added to one or more of the lists (I haven't included a list for 'no list' because although there is never enough, 10 is enough for this little thing)
- View all channels in each list
- Pressing and holding `*Scan` will start the scanning.
  - If any lists previously already enabled will be returned
  - If no lists are enabled, it stays on the first channel
  - It scans through all enabled list(s) and channel(s) in numerical order
  - Pressing any of the numerical keys `0`,`1`,`2`,`3`,`4`,`5`,`6`,`7`,`8`,`9` turns the respective list on or off
    - Turning a list on flashes the light green
    - Turning a list off flashes the light red
<br/><br/>
The video below shows it scanning lists 1,3,5 (I slowed it down for demonstration by giving a delay in the code)
Key:
`-` = list disabled
`*` = list enabled but not currently being scanned
`0-9` = list enabled and currently being scanned

https://github.com/egzumer/uv-k5-firmware-custom/assets/68870168/20675ff8-755d-4f86-98b6-b076a1df91c1

<br/><br/>
These photos show the lists for one channel 

![PXL_20240219_121952876-ScanAdd0](https://github.com/egzumer/uv-k5-firmware-custom/assets/68870168/3bb73d9e-1c74-46f2-ace1-74b44642aaf1)
![PXL_20240219_121958761-ScanAdd3](https://github.com/egzumer/uv-k5-firmware-custom/assets/68870168/a58f1f74-c2ee-474f-af8f-fe58a18aed44)
![PXL_20240219_122008594-ScanAdd05](https://github.com/egzumer/uv-k5-firmware-custom/assets/68870168/844b220b-6eb4-4a7b-9c40-d3619de16289)
<br/><br/>
The video below shows going through the lists and viewing the channels in each list.  Note: Channel 21 is in the lists 0,2,4,6,7,8,9 because I wanted something in them while testing.

https://github.com/egzumer/uv-k5-firmware-custom/assets/68870168/700be171-3a25-4b0d-ad73-53615ba3fde0

**Comment by qsus** [comment]
As I already said, I would love to see this in the egzumer firmware.
However, if I understand correctly, this feature takes up about 20 entries in the menu. I see you made it possible to add a channel to more lists, so it isn't possible to just make one entry with value 0-9. The same thing about listing the channels. Perhaps the best way would be to create a separate lists menu, perhaps using holding 5 key, which was now used to change lists for a given channel?
I personally don't need managing lists in the radio, because I set lists inside Chirp; but if this becomes part of the firmware, it shouldn't fill the whole menu.

**Reply by AubsUK** [reply]
You're right @qsus it does take up 20 menu options (10 for adding, 10 for listing).  The original code for 2 lists took up 5 (2 for adding, 2 for listing, 1 for active lists).

I did wonder if the existing code has capability for nested menus, but I haven't seen any currently present in the radio, so assume not, and it might take a lot of rewriting of original code to get that functionality.  Perhaps @egzumer might have some knowledge on that?

**Reply by egzumer** [reply]
There is no menu nesting.

One of the reasons I started as a fork of 1o11 because I didn't wan't more changes to the firmware that I didn't need. I'm sorry to say, but I have to say no to those changes. I won't ever use that. 2 lists is more than I need.

I already added a note to promote other forks, I'm very much open to expand on that. I could create one place where everyone could post a link and a description of their changes to make it easy for everyone to find what they need.
That way I don't have to spend time on that, hunt new bugs, coordinate anything.

**Reply by AubsUK** [reply]
@qsus - I've just finished the firmware and CHIRP driver over on my repo - https://github.com/AubsUK/uv-k5-firmware-custom

Hope you like it.  Open a discussion there if you have any suggestions.

**Comment by ErikS-web** [comment]
Nice experiment, but ....

Are you planning to use so much scanlists as _portable radio_, or at home ?

You might consider using [Q-Dock](https://github.com/nicsure/QuanshengDock/wiki/50-%E2%80%90-Specific-Functions-explained#scanning-ranges-extended-configuration)

Than you can make a much scanlists as you wish (on the PC), even without the use of Chirp.

**Reply by AubsUK** [reply]
Think about taking your HT with you as you regularly travel.  I know you can just program in the 'standard' frequencies for repeaters, but if you go to different areas regularly, having different scan lists set up speeds up only scanning the repeaters you want.

Wherever you might be (home or mobile), being able to scan 2m or 70cm repeaters, 2m or 70cm simplex independently, is always a good thing.  Too much chatter on 70cm repeaters? - Turn that list off and continue to scan 2m repeaters and 2m/70cm simplex.

Same for things like PMR, Marine VHF, CB, Air.

Scanning everything (say all channels) or just one list (say half channels) does take a while, so when not near the coast/rivers/canals, the only way would be to remove the Marine channels from the list manually, re-program on Chirp or Q-Dock.  Same for when you're out and about you might not want to scan air.

Definitely each to their own, but when you have the capability to not only TX on specific bands, but to also to RX other frequencies when you want, why not use it.
